### PR TITLE
Move `doc(primitive)` future incompat warning to `invalid_doc_attributes`

### LINKED ISSUE #88070: Tracking Issue for `#![feature(doc_primitive)]`
This is a tracking issue for using `#[doc(primitive)]`.
The feature gate for the issue is `#![feature(doc_primitive)]`.

This feature is not meant to ever become stable. It is an implementation detail of the standard library.

Due to an oversight when releasing Rust 1.0, this could previously be used without a feature gate. That was a bug: this feature is not useful to anyone other than the standard library, and the behavior when multiple crates use `doc(primitive)` at once is unspecified (see "Breaking changes to doc(primitive)" in https://github.com/rust-lang/rust/pull/87073). To avoid needlessly breaking crates, anyone using `doc(primitive)` without a feature gate will get a future-incompat warning instead of a hard error.

`#[doc(primitive)]` still has an effect even without the feature gate, but the behavior is still unspecified when more than one crate has the same `primitive` attribute. Once https://github.com/rust-lang/rust/pull/87073 is merged and `core` starts documenting all primitives, it will be completely unspecified whether the new documentation is used or not.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Initial implementation: https://github.com/rust-lang/rust/pull/87050
- [ ] Make this a hard error

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

Does this need an edition to become a hard error? Crater hasn't found any crates using this attribute: https://github.com/rust-lang/rust/pull/87050#issuecomment-886166706

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #88070.

It's been a while since this was turned into a "future incompatible lint" so I think we can now turn it into a hard error without problem.

r? @jyn514

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Good idea!

**Comment by GuillaumeGomez** [comment]
To make it work, I needed to update how it's handled in rustdoc and to put it behind a `cfg` in the `std`.

**Comment by jyn514** [comment]
I don't have time for reviews right now.

r? rustdoc cc @petrochenkov

**Comment by GuillaumeGomez** [comment]
cc @petrochenkov since they worked on something similar in https://github.com/rust-lang/rust/pull/109534.

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -264,18 +264,13 @@ impl ExternalCrate {
         let as_primitive = |res: Res<!>| {
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:264]: It's much better, thanks!

**Comment by jyn514** [comment]
Shouldn't we do a crater run for this change?

**Comment by jyn514** [comment]
Ok, please update the PR description to reflect what the PR now does.

**Comment by GuillaumeGomez** [comment]
@jyn514 Considering we have a warning that have been around for quite a while, I think it's ok. But do you see a reason where it could a be a problem maybe?

**Comment by GuillaumeGomez** [comment]
Found the problem: since some parts of intra-doc link generation was moved into `rustc_resolve`, it used `doc(primitive)` in there too but wasn't updated.

**Pull request merged**